File: unity-client/Assets/Editor/KromkaWalkCollision.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using UnityEngine;

namespace Kromka.EditorTools
{
    /// <summary>
    /// Какая часть коллайдеров сцены останавливает идущего — так, как это видит сервер:
    /// плоские фигуры на земле. Клиент останавливает игрока самими коллайдерами, сервер —
    /// строкой data, поэтому экспорт пишет эти фигуры в collisionParts, а
    /// src/server/location-collision.js собирает из них преграды.
    ///
    /// Зеркало tools/kromka-walk-collision.js: tools/check-kromka-collision-parity.js
    /// пересчитывает те же фигуры из YAML сцены и сверяет с data, поэтому любая правка
    /// правил здесь обязана повториться там.
    /// </summary>
    internal static class KromkaWalkCollision
    {
        // Игрок — капсула 1,8 м с шагом 0,25 м (RoaPlayerController) на земле, верх которой
        // лежит на y = -0,05 (Ground_EDITABLE). Считается только то, что у коллайдера есть
        // между шагом и головой: под крышей, балкой крана и баком на опорах проходят, а
        // наклонный конвейер мешает лишь там, где лента опускается до роста человека.
        internal const double BandBottom = 0.2;
        internal const double BandTop = 1.75;
        // Полоса тоньше — касание (кромка панели на 5 мм ниже головы), а не преграда.
        internal const double MinPartSize = 0.05;
        private const double Upright = 0.999;
        private const double Horizontal = 0.001;

        internal struct GroundShape
        {
            public bool Round;
            public double X, Z, Radius, SizeX, SizeZ, Yaw;
        }

        /// <summary>Включённые коллайдеры объекта, которые физика действительно учитывает.</summary>
        internal static List<Collider> SolidColliders(GameObject root)
        {
            var result = new List<Collider>();
            if (root == null) return result;
            foreach (Collider collider in root.GetComponentsInChildren<Collider>(true))
                if (collider != null && collider.enabled && !collider.isTrigger && collider.gameObject.activeInHierarchy)
                    result.Add(collider);
            return result;
        }

        /// <summary>Останавливает ли объект идущего: есть ли у него хоть что-то на высоте тела.</summary>
        internal static bool BlocksWalking(GameObject root) => Shapes(root).Count > 0;

        internal static List<GroundShape> Shapes(GameObject root)
        {
            var shapes = new List<GroundShape>();
            foreach (Collider collider in SolidColliders(root))
            {
                GroundShape? shape = ShapeOf(collider);
                if (shape == null) continue;
                GroundShape value = shape.Value;
                if (value.Round ? value.Radius * 2 < MinPartSize : Math.Min(value.SizeX, value.SizeZ) < MinPartSize)
                    continue;
                shapes.Add(value);
            }
            return shapes;
        }

        /// <summary>
        /// collisionParts строки в той системе, в которой их читает сервер: смещение от
        /// позиции строки, развёрнутое обратно на её поворот и делённое на её масштаб.
        /// Позиция, поворот и масштаб берутся уже округлёнными — такими их увидит сервер.
        /// </summary>
        internal static JArray Parts(List<GroundShape> shapes, Vector3 position, double yaw, Vector3 scale)
        {
            double scaleX = Math.Abs(scale.x) > 1e-6 ? scale.x : 1;
            double scaleZ = Math.Abs(scale.z) > 1e-6 ? scale.z : 1;
            double cos = Math.Cos(yaw);
            double sin = Math.Sin(yaw);
            var parts = new JArray();
            foreach (GroundShape shape in shapes)
            {
                double dx = shape.X - position.x;
                double dz = shape.Z - position.z;
                var part = new JObject
                {
                    ["center"] = new JObject
                    {
                        ["x"] = Round((dx * cos - dz * sin) / scaleX),
                        ["z"] = Round((dx * sin + dz * cos) / scaleZ)
                    }
                };
                if (shape.Round)
                {
                    part["radius"] = Round(shape.Radius / Math.Max(Math.Abs(scaleX), Math.Abs(scaleZ)));
                }
                else
                {
                    part["size"] = new JObject
                    {
                        ["x"] = Round(shape.SizeX / Math.Abs(scaleX)),
                        ["z"] = Round(shape.SizeZ / Math.Abs(scaleZ))
                    };
                    double turned = Round(HalfTurn(shape.Yaw - yaw));
                    if (turned != 0) part["rotationY"] = turned;
                }
                parts.Add(part);
            }
            return parts;
        }

        /// <summary>
        /// Размах включённых коллайдеров по земле — footprint строки. Считается из самих
        /// фигур, а не из Collider.bounds, чтобы проверка данных повторяла его без Unity.
        /// </summary>
        internal static bool TryFootprint(GameObject root, out double sizeX, out double sizeZ)
        {
            double minX = double.PositiveInfinity, maxX = double.NegativeInfinity;
            double minZ = double.PositiveInfinity, maxZ = double.NegativeInfinity;
            foreach (Collider collider in SolidColliders(root))
            {
                foreach (Vector3 point in ExtremePoints(collider, out double pad))
                {
                    minX = Math.Min(minX, point.x - pad); maxX = Math.Max(maxX, point.x + pad);
                    minZ = Math.Min(minZ, point.z - pad); maxZ = Math.Max(maxZ, point.z + pad);
                }
            }
            sizeX = maxX - minX;
            sizeZ = maxZ - minZ;
            return !double.IsInfinity(minX);
        }

        private static IEnumerable<Vector3> ExtremePoints(Collider collider, out double pad)
        {
            Transform transform = collider.transform;
            pad = 0;
            if (collider is BoxCollider box) return BoxCorners(transform, box.center, box.size * 0.5f);
            if (collider is SphereCollider sphere)
            {
                pad = sphere.radius * MaxAbs(transform.lossyScale);
                return new[] { transform.TransformPoint(sphere.center) };
            }
            if (collider is CapsuleCollider capsule)
            {
                CapsuleWorld(capsule, out Vector3 center, out Vector3 axis, out double radius, out double straight);
                pad = radius;
                return new[] { center + axis * (float)straight, center - axis * (float)straight };
            }
            throw Unsupported(collider);
        }

        private static GroundShape? ShapeOf(Collider collider)
        {
            Transform transform = collider.transform;
            if (collider is BoxCollider box)
                return BoxShape(BoxCorners(transform, box.center, box.size * 0.5f), transform);

            if (collider is SphereCollider sphere)
            {
                Vector3 center = transform.TransformPoint(sphere.center);
                double radius = sphere.radius * MaxAbs(transform.lossyScale);
                double away = DistanceToBand(center.y, center.y);
                if (away >= radius) return null;
                return new GroundShape { Round = true, X = center.x, Z = center.z, Radius = Math.Sqrt(radius * radius - away * away) };
            }

            if (collider is CapsuleCollider capsule)
            {
                CapsuleWorld(capsule, out Vector3 center, out Vector3 axis, out double radius, out double straight);
                if (Math.Abs(axis.y) > Upright)
                {
                    double away = DistanceToBand(center.y - straight, center.y + straight);
                    if (away >= radius) return null;
                    return new GroundShape { Round = true, X = center.x, Z = center.z, Radius = Math.Sqrt(radius * radius - away * away) };
                }
                if (Math.Abs(axis.y) < Horizontal)
                {
                    // Труба: шириной с хорду круглого сечения на высоте тела, длиной с прямой
                    // участок; скруглённые торцы добавляют сторону вписанного в них квадрата.
                    double away = DistanceToBand(center.y, center.y);
                    if (away >= radius) return null;
                    double chord = Math.Sqrt(radius * radius - away * away);
                    double length = Math.Sqrt((double)axis.x * axis.x + (double)axis.z * axis.z);
                    if (length < 1e-9) length = 1;
                    return new GroundShape
                    {
                        X = center.x,
                        Z = center.z,
                        SizeX = (straight + chord * Math.Sqrt(0.5)) * 2,
                        SizeZ = chord * 2,
                        Yaw = YawOf(axis.x / length, axis.z / length)
                    };
                }
                // У наклонной капсулы простой фигуры нет; её габаритный ящик ошибается в сторону преграды.
                Vector3 half = Vector3.one * capsule.radius;
                half[capsule.direction] = Mathf.Max(capsule.height * 0.5f, capsule.radius);
                return BoxShape(BoxCorners(transform, capsule.center, half), transform);
            }

            throw Unsupported(collider);
        }

        private static void CapsuleWorld(CapsuleCollider capsule, out Vector3 center, out Vector3 axis,
                                         out double radius, out double straight)
        {
            Transform transform = capsule.transform;
            Vector3 scale = transform.lossyScale;
            int direction = capsule.direction;
            int a = (direction + 1) % 3;
            int b = (direction + 2) % 3;
            center = transform.TransformPoint(capsule.center);
            axis = direction == 0 ? transform.right : direction == 1 ? transform.up : transform.forward;
            radius = capsule.radius * Math.Max(Math.Abs(scale[a]), Math.Abs(scale[b]));
            straight = Math.Max(0, capsule.height * Math.Abs(scale[direction]) * 0.5 - radius);
        }

        private static Vector3[] BoxCorners(Transform transform, Vector3 center, Vector3 half)
        {
            var corners = new Vector3[8];
            for (int i = 0; i < 8; i++)
                corners[i] = transform.TransformPoint(center + new Vector3(
                    (i & 1) != 0 ? half.x : -half.x, (i & 2) != 0 ? half.y : -half.y, (i & 4) != 0 ? half.z : -half.z));
            return corners;
        }

        // Срез ящика полосой высот тела, спроецированный на землю и обведённый прямоугольником.
        private static GroundShape? BoxShape(Vector3[] corners, Transform transform)
        {
            var points = new List<double[]>();
            foreach (Vector3 corner in corners)
                if (corner.y >= BandBottom && corner.y <= BandTop) points.Add(new double[] { corner.x, corner.z });
            for (int a = 0; a < 8; a++)
            {
                foreach (int bit in new[] { 1, 2, 4 })
                {
                    if ((a & bit) != 0) continue;
                    Vector3 from = corners[a];
                    Vector3 to = corners[a | bit];
                    foreach (double level in new[] { BandBottom, BandTop })
                    {
                        if ((from.y - level) * (to.y - level) >= 0) continue;
                        double t = (level - from.y) / ((double)to.y - from.y);
                        points.Add(new[] { from.x + (to.x - from.x) * t, from.z + (to.z - from.z) * t });
                    }
                }
            }
            if (points.Count == 0) return null;

            // Рамка идёт по самой горизонтальной оси ящика: стоячий или только наклонённый
            // ящик сохраняет собственный прямоугольник, а не больший по осям мира.
            Vector3[] axes = { transform.right, transform.up, transform.forward };
            int level0 = 0;
            for (int index = 1; index < 3; index++)
                if (Math.Abs(axes[index].y) < Math.Abs(axes[level0].y) - 1e-6) level0 = index;
            double length = Math.Sqrt((double)axes[level0].x * axes[level0].x + (double)axes[level0].z * axes[level0].z);
            if (length < 1e-9) length = 1;
            double yaw = YawOf(axes[level0].x / length, axes[level0].z / length);

            double cos = Math.Cos(yaw);
            double sin = Math.Sin(yaw);
            double minU = double.PositiveInfinity, maxU = double.NegativeInfinity;
            double minV = double.PositiveInfinity, maxV = double.NegativeInfinity;
            foreach (double[] point in points)
            {
                double u = point[0] * cos - point[1] * sin;
                double v = point[0] * sin + point[1] * cos;
                minU = Math.Min(minU, u); maxU = Math.Max(maxU, u);
                minV = Math.Min(minV, v); maxV = Math.Max(maxV, v);
            }
            double centerU = (minU + maxU) * 0.5;
            double centerV = (minV + maxV) * 0.5;
            return new GroundShape
            {
                X = centerU * cos + centerV * sin,
                Z = -centerU * sin + centerV * cos,
                SizeX = maxU - minU,
                SizeZ = maxV - minV,
                Yaw = yaw
            };
        }

        // Поворот Unity вокруг Y, при котором локальная ось X смотрит вдоль (x, z) по земле.
        private static double YawOf(double x, double z) => Math.Atan2(-z, x);

        // Прямоугольник, повёрнутый на пол-оборота, — тот же прямоугольник.
        private static double HalfTurn(double angle)
        {
            double value = angle % Math.PI;
            if (value > Math.PI / 2) value -= Math.PI;
            if (value <= -Math.PI / 2) value += Math.PI;
            return value;
        }

        private static double DistanceToBand(double low, double high)
        {
            if (high < BandBottom) return BandBottom - high;
            if (low > BandTop) return low - BandTop;
            return 0;
        }

        private static double MaxAbs(Vector3 value) =>
            Math.Max(Math.Abs(value.x), Math.Max(Math.Abs(value.y), Math.Abs(value.z)));

        internal static double Round(double value)
        {
            double rounded = Math.Round(value, 3, MidpointRounding.AwayFromZero);
            return rounded == 0 ? 0 : rounded;
        }

        private static Exception Unsupported(Collider collider) => new InvalidOperationException(
            "Коллайдер " + collider.GetType().Name + " объекта " + collider.name
            + " не имеет серверной формы: сцены Кромки собираются из Box, Sphere и Capsule.");
    }
}

File: tools/kromka-walk-collision.js
'use strict';

// Which part of a Unity scene collider stops a walking body, as the server sees it:
// a 2D shape on the ground plane. Mirrors KromkaWalkCollision in
// unity-client/Assets/Editor/KromkaWorldSceneExporter.cs, which writes these shapes
// into `collisionParts`; tools/check-kromka-collision-parity.js recomputes them from
// the scene YAML and compares, so a scene and its data cannot drift apart.
//
// A player is a 1.8 m capsule that steps over 0.25 m (RoaPlayerController) on ground
// whose top lies at y = -0.05 (Ground_EDITABLE). Only what a collider has between the
// step and the head counts: a roof, a gantry beam or a water tank on legs is walked
// under, and a slanted conveyor blocks only where its belt comes down to head height.

const WALK_BAND_BOTTOM = 0.2;
const WALK_BAND_TOP = 1.75;
// A sliver thinner than this is a grazing contact (a panel edge 5 mm below head
// height), not an obstacle the server should turn into a wall.
const MIN_PART_SIZE = 0.05;
const UPRIGHT = 0.999;
const HORIZONTAL = 0.001;

const BOX_EDGES = [];
for (let a = 0; a < 8; a++) for (const bit of [1, 2, 4]) if (!(a & bit)) BOX_EDGES.push([a, a | bit]);

// Half away from zero, like Math.Round(value, 3, MidpointRounding.AwayFromZero) in the exporter.
function round3(value) {
  const rounded = Math.sign(value) * Math.round(Math.abs(value) * 1000) / 1000;
  return rounded === 0 ? 0 : rounded;
}

// Unity yaw of a frame whose local X points along the ground direction (x, z).
function yawOf(x, z) {
  return Math.atan2(-z, x);
}

// A rectangle turned by half a turn is the same rectangle.
function halfTurn(angle) {
  let value = angle % Math.PI;
  if (value > Math.PI / 2) value -= Math.PI;
  if (value <= -Math.PI / 2) value += Math.PI;
  return value;
}

function distanceToBand(low, high) {
  if (high < WALK_BAND_BOTTOM) return WALK_BAND_BOTTOM - high;
  if (low > WALK_BAND_TOP) return low - WALK_BAND_TOP;
  return 0;
}

// The tightest rectangle, turned by `yaw`, around ground points.
function rectangleAround(points, yaw) {
  const cos = Math.cos(yaw);
  const sin = Math.sin(yaw);
  let minU = Infinity, maxU = -Infinity, minV = Infinity, maxV = -Infinity;
  for (const [x, z] of points) {
    const u = x * cos - z * sin;
    const v = x * sin + z * cos;
    minU = Math.min(minU, u); maxU = Math.max(maxU, u);
    minV = Math.min(minV, v); maxV = Math.max(maxV, v);
  }
  const u = (minU + maxU) / 2;
  const v = (minV + maxV) / 2;
  return { x: u * cos + v * sin, z: -u * sin + v * cos, sizeX: maxU - minU, sizeZ: maxV - minV, yaw };
}

function boxShape(corners, axes) {
  const points = [];
  for (const corner of corners)
    if (corner[1] >= WALK_BAND_BOTTOM && corner[1] <= WALK_BAND_TOP) points.push([corner[0], corner[2]]);
  for (const [a, b] of BOX_EDGES) {
    const from = corners[a];
    const to = corners[b];
    for (const level of [WALK_BAND_BOTTOM, WALK_BAND_TOP]) {
      if ((from[1] - level) * (to[1] - level) >= 0) continue;
      const t = (level - from[1]) / (to[1] - from[1]);
      points.push([from[0] + (to[0] - from[0]) * t, from[2] + (to[2] - from[2]) * t]);
    }
  }
  if (!points.length) return null;
  // The frame follows the box's most level axis, so an upright or merely pitched box
  // keeps its own rectangle instead of a larger world-aligned one.
  let level = 0;
  for (let index = 1; index < 3; index++)
    if (Math.abs(axes[index][1]) < Math.abs(axes[level][1]) - 1e-6) level = index;
  const length = Math.hypot(axes[level][0], axes[level][2]) || 1;
  return rectangleAround(points, yawOf(axes[level][0] / length, axes[level][2] / length));
}

function boxCornersOfCapsule(shape) {
  const half = [shape.radius, shape.radius, shape.radius];
  half[shape.direction] = Math.max(shape.height / 2, shape.radius);
  const corners = [];
  for (let i = 0; i < 8; i++) {
    const local = [(i & 1) ? half[0] : -half[0], (i & 2) ? half[1] : -half[1], (i & 4) ? half[2] : -half[2]];
    corners.push([0, 1, 2].map(row => shape.center[row]
      + shape.axes[0][row] * local[0] * shape.scale[0]
      + shape.axes[1][row] * local[1] * shape.scale[1]
      + shape.axes[2][row] * local[2] * shape.scale[2]));
  }
  return corners;
}

// World-space ground shapes of one collider: { x, z, radius } or { x, z, sizeX, sizeZ, yaw }.
function colliderShapes(shape) {
  if (shape.kind === 'box') {
    const rectangle = boxShape(shape.corners, shape.axes);
    return rectangle ? [rectangle] : [];
  }
  if (shape.kind === 'sphere') {
    const radius = shape.radius * Math.max(...shape.scale);
    const away = distanceToBand(shape.center[1], shape.center[1]);
    if (away >= radius) return [];
    return [{ x: shape.center[0], z: shape.center[2], radius: Math.sqrt(radius * radius - away * away) }];
  }
  if (shape.kind !== 'capsule') throw new Error(`no ground shape for a ${shape.kind} collider`);
  const axis = shape.axes[shape.direction];
  const around = [0, 1, 2].filter(index => index !== shape.direction);
  const radius = shape.radius * Math.max(shape.scale[around[0]], shape.scale[around[1]]);
  const straight = Math.max(0, shape.height * shape.scale[shape.direction] / 2 - radius);
  if (Math.abs(axis[1]) > UPRIGHT) {
    const away = distanceToBand(shape.center[1] - straight, shape.center[1] + straight);
    if (away >= radius) return [];
    return [{ x: shape.center[0], z: shape.center[2], radius: Math.sqrt(radius * radius - away * away) }];
  }
  if (Math.abs(axis[1]) < HORIZONTAL) {
    // A pipe: as wide as the chord of its round section at body height, as long as its
    // straight run; the rounded ends add the side of the square inscribed in them.
    const away = distanceToBand(shape.center[1], shape.center[1]);
    if (away >= radius) return [];
    const chord = Math.sqrt(radius * radius - away * away);
    const length = Math.hypot(axis[0], axis[2]) || 1;
    return [{
      x: shape.center[0],
      z: shape.center[2],
      sizeX: (straight + chord * Math.SQRT1_2) * 2,
      sizeZ: chord * 2,
      yaw: yawOf(axis[0] / length, axis[2] / length)
    }];
  }
  // A leaning capsule has no simple ground shape; its bounding box errs on the blocking side.
  const rectangle = boxShape(boxCornersOfCapsule(shape), shape.axes);
  return rectangle ? [rectangle] : [];
}

// The collisionParts of a row, in the frame the server decodes them in
// (src/server/location-collision.js): offset from the row position, turned back by
// the row's yaw, divided by the row's scale.
function walkCollisionParts(colliders, row) {
  const position = row.position || {};
  const yaw = Number(row.rotation?.y || 0);
  const scaleX = Math.abs(Number(row.scale?.x)) > 1e-6 ? Number(row.scale.x) : 1;
  const scaleZ = Math.abs(Number(row.scale?.z)) > 1e-6 ? Number(row.scale.z) : 1;
  const cos = Math.cos(yaw);
  const sin = Math.sin(yaw);
  const parts = [];
  for (const collider of colliders) {
    for (const shape of colliderShapes(collider)) {
      if (shape.radius !== undefined ? shape.radius * 2 < MIN_PART_SIZE
        : Math.min(shape.sizeX, shape.sizeZ) < MIN_PART_SIZE) continue;
      const dx = shape.x - Number(position.x || 0);
      const dz = shape.z - Number(position.z || 0);
      const part = { center: { x: round3((dx * cos - dz * sin) / scaleX), z: round3((dx * sin + dz * cos) / scaleZ) } };
      if (shape.radius !== undefined) part.radius = round3(shape.radius / Math.max(Math.abs(scaleX), Math.abs(scaleZ)));
      else {
        part.size = { x: round3(shape.sizeX / Math.abs(scaleX)), z: round3(shape.sizeZ / Math.abs(scaleZ)) };
        const turned = round3(halfTurn(shape.yaw - yaw));
        if (turned !== 0) part.rotationY = turned;
      }
      parts.push(part);
    }
  }
  return parts;
}

// The ground extent of every solid collider of a row, overhead parts included: its
// `footprint` (the minimap draws it; the server falls back to it only without parts).
function colliderFootprint(colliders) {
  let minX = Infinity, maxX = -Infinity, minZ = Infinity, maxZ = -Infinity;
  const take = (x, z, pad) => {
    minX = Math.min(minX, x - pad); maxX = Math.max(maxX, x + pad);
    minZ = Math.min(minZ, z - pad); maxZ = Math.max(maxZ, z + pad);
  };
  for (const shape of colliders) {
    if (shape.kind === 'box') for (const corner of shape.corners) take(corner[0], corner[2], 0);
    else if (shape.kind === 'sphere') take(shape.center[0], shape.center[2], shape.radius * Math.max(...shape.scale));
    else {
      const axis = shape.axes[shape.direction];
      const around = [0, 1, 2].filter(index => index !== shape.direction);
      const radius = shape.radius * Math.max(shape.scale[around[0]], shape.scale[around[1]]);
      const straight = Math.max(0, shape.height * shape.scale[shape.direction] / 2 - radius);
      for (const side of [1, -1]) take(shape.center[0] + axis[0] * straight * side, shape.center[2] + axis[2] * straight * side, radius);
    }
  }
  if (minX === Infinity) return null;
  return { x: round3(Math.max(0.2, maxX - minX)), z: round3(Math.max(0.2, maxZ - minZ)) };
}

module.exports = {
  MIN_PART_SIZE,
  WALK_BAND_BOTTOM,
  WALK_BAND_TOP,
  colliderFootprint,
  colliderShapes,
  walkCollisionParts
};
